### PR TITLE
Index now supports multiple relations with the same name.

### PR DESCRIPTION
**Pull request by dhensby**
Replaces #144

### CODE REVIEW

File: code/search/SearchIndex.php [hunk 1]
```diff
@@ -37,6 +37,27 @@
...
+     */
+    private static $class_delimiter = '_|_';
+
+    /**
+     * This is used to clean the source name from suffix
+     * suffixes are needed to support multiple relations with the same name on different page types
+     * @param string $source
+     * @return string
+     */
+    private function getSourceName($source)
```
Comment (chillu) [inline code/search/SearchIndex.php]: We don't make functions private usually, please mark it protected unless there's a really good reason that a subclass shouldn't be able to alter this behaviour

File: code/search/SearchIndex.php [hunk 2]
```diff
@@ -77,26 +98,49 @@ public function fieldData($field, $forceType = null, $extraOptions = array())
...
+                    $source = $this->getSourceName($source);
 
-                    foreach (SearchIntrospection::hierarchy($source, $options['include_children']) as $dataclass) {
+                    foreach (SearchIntrospection::hierarchy($source, $baseOptions['include_children']) as $dataclass) {
+                        $class = null;
+                        $options = $baseOptions;
                         $singleton = singleton($dataclass);
 
                         if ($hasOne = $singleton->has_one($lookup)) {
+                            // we only want to include base class for relation, omit classes that inherited the relation
```
Comment (chillu) [inline code/search/SearchIndex.php]: Who do we exclude inherited relations here? That seems like a recipe for a lot of dev confusion. From my understanding, we do include inherited fields, e.g. if you use `$myIndex->addClass('Page')` and `$myIndex->addField('Content')`, it'll index `Content` even though it's defined on the parent class (`SiteTree`)

File: code/search/SearchIndex.php [hunk 3]
```diff
@@ -77,26 +98,49 @@ public function fieldData($field, $forceType = null, $extraOptions = array())
             foreach ($lookups as $lookup) {
                 $next = array();
 
-                foreach ($sources as $source => $options) {
-                    $class = null;
+                foreach ($sources as $source => $baseOptions) {
```
Comment (chillu) [inline code/search/SearchIndex.php]: Please update the `fieldData()` method docs to note on how it handles relationships

File: code/search/SearchIndex.php (hunk 2 above)
Comment (mfendeksilverstripe) [inline code/search/SearchIndex.php]: I don't think there is a functional difference between adding the inherited relations or not as the index will fall back on the parent class relation. The difference would be the content of the XML which is output of this process. We probably don't want to have the XML full of fields that have no impact on the search.

Of course I'm happy to discuss this further as I'm quite new to this module.

**Comment by mfendeksilverstripe** [comment]
@dhensby Can you please give me write access to this fork? I would like to add some changes based on feedback from Ingo.

**Comment by mfendeksilverstripe** [comment]
@dhensby @chillu I processed the feedback and made some new changes. I also includes Ingo's tests and my own as well. I pushed the new version to my own fork because I don't have access to this one. I also addressed one of the issues Ingo raised (include classes that inherit the relation). Such classes do not need to be included in the index because their parent class will be included instead (regardless if parent class was originally meant to be included in the index or not). This is because the inner loop calls `hierarchy` which will include the parent class if needed.

https://github.com/mfendeksilverstripe/silverstripe-fulltextsearch/commit/32a6cfa8030ef7c091f1880fe1a7ff510f8923ab

Please have a look and let me know if you have more feedback for me.

**Comment by dhensby** [comment]
I've integrated these changes.

If you need to make further changes either @chillu can push to my branch (he has permissions) or feel free to close this PR and open a new one. I'm not sure how to add permissions to individual branches to specific users.

**Pull request merged**